### PR TITLE
Support for inlay hints

### LINKED ISSUE #473: Feature request- Support for LSP inlay hints?
Hey 👋 
Thanks for your great work here!
I was wondering if there is any plan or agenda to implement the inlay hint provider of the language server protocol. I find them extremely useful in other languages and I’m missing them quite often while writing Kotlin. 
a repository search for `inlay` left no results, so I just open this issue. 
Thanks for any insights.

### PR DESCRIPTION
**Pull request by ElamC**
Closes  #473 

Adds inlay hint support for inferred types and parameter names. 

Todo:
- [x] add tests
- [x] hints for lambdas

~~consider adding a resolve method for [versioning](https://github.com/rust-lang/rust-analyzer/pull/13697) to solve textDocument/inlayHint request failure~~

### CODE REVIEW

**Comment by ElamC** [comment]
> Did some manual testing, and it seems to work really great 🙂 Awesome work! 👍 Looking forward to seeing the tests, which I see in your todo-list.

Added tests, let me know how it looks

File: server/src/main/kotlin/org/javacs/kt/inlayhints/InlayHint.kt
Comment (themkat) [inline server/src/main/kotlin/org/javacs/kt/inlayhints/InlayHint.kt:87]: You get a detekt (static code analysis warning) in your build for the nesting depth of the function here here. Maybe the inner label creation (line 92) could be extracted to its own helper function? 🙂

File: server/src/main/kotlin/org/javacs/kt/inlayhints/InlayHint.kt
Comment (themkat) [inline server/src/main/kotlin/org/javacs/kt/inlayhints/InlayHint.kt:54]: TODOs are not allowed in detekt. If this is something that should be looked into further (i.e, after this PR is merged), I suggest creating an issue for it. Comments with TODOs have a tendency to become stale after a while, and sometimes people forget to remove them.

File: server/src/test/kotlin/org/javacs/kt/InlayHintTest.kt
Comment (themkat) [inline server/src/test/kotlin/org/javacs/kt/InlayHintTest.kt:133]: I never use inlay hints, and am not used to it in IntelliJ (I stay away from big IDEs like the plague), so my question may be stupid. Is this specific inlay hint for the intermediate map operation on line 19 in the ChainedMethods.kt test file?
```kotlin
val bar = listOf("hello", "world")
    .map { it.length * 2 }
```

I did not think there was supposed to be anything in those cases. If it makes sense, then it is probably okay 🙂 Might just also be me being tired after work and trying to understand stuff before dinner 😆

**Review by themkat — COMMENTED** [review]
Looks good to me! 😄 Some minor comments below. Some of the detekt issues will be gone by itself once #513 is merged (once rebased/merged into your branch ofc), but the ones that will not I have commented below.

Pretty sure this is ready for merge after that is done. Will have a final look and approve at that time 🙂 Also to give @fwcd time to have a look in case they want to.

File: server/src/test/kotlin/org/javacs/kt/InlayHintTest.kt
Comment (ElamC) [reply server/src/test/kotlin/org/javacs/kt/InlayHintTest.kt:133]: Yup, thats the hint for `.map {}`. This is how it looks like with intellij's default hints:

Screenshot 2023-10-11 at 20 59 53

**Comment by ElamC** [comment]
@themkat is there a way we can skip some of the issues detekt raises?

**Comment by themkat** [comment]
> @themkat @fwcd is there a way we can skip some of the issues detekt raises?

In general, one can use the `@Suppress` annotation for that (e.g, `@Suppress("EmtpyClassBlock")`) 🙂 Then you can skip rules in general places. Disabling rules for the entires code base is done in `detekt.yml` in the root-directory. 

I see you have fixed the issues while waiting for our reply. Good job! Sorry for being a bit slow, have been busy in the last few days 🙁

**Comment by themkat** [comment]
Found one case we might consider adding. It is okay if you don't add it in this PR, but then we should create an issue for it 🙂 

The case is the one-line function syntax that Kotlin support. Notice the function `myfunc` below:
image

We see that type hints appears as they should for the variable, but nothing for the function. Unsure if other IDEs give type hints in this case, but in my view it would probably be neat to have 🙂

**Review by themkat — APPROVED** [review]
Looks very good to me! 🙂 Thanks again for fixing the detekt issues. 

Also see my previous comment. I'm open to merging this PR without it, but if it is reasonable we should create an issue for it.

**Comment by ElamC** [comment]
> Found one case we might consider adding. It is okay if you don't add it in this PR, but then we should create an issue for it 🙂
> 
> The case is the one-line function syntax that Kotlin support. Notice the function `myfunc` below: image
> 
> We see that type hints appears as they should for the variable, but nothing for the function. Unsure if other IDEs give type hints in this case, but in my view it would probably be neat to have 🙂

good catch, I can add it to this PR. 
Also would be nice to have additional control over settings i.e. toggling different hints since now all types are rendered by default. This is something I can work on for this PR too.

**Comment by ElamC** [comment]
@themkat let me know what you think but I believe we can merge this now. 
Just an FYI - to reflect config changes the vscode plugin needs to be updated (open an issue for this?). My setup using neovim looks like this:

```lua
lspconfig.kotlin_language_server.setup {
    cmd = { ... },
    capabilities = capabilities,
    on_attach = on_attach,
    settings = {
        kotlin = {
            hints = {
                typeHints = true,
                parameterHints = true,
                chainedHints = false,
            },
        },
    },
}
```

File: server/src/main/kotlin/org/javacs/kt/inlayhints/InlayHint.kt
Comment (themkat) [inline server/src/main/kotlin/org/javacs/kt/inlayhints/InlayHint.kt:198]: Why does this not have any tests? I know the code is small, but tests helps us not regress. It is also easier to extend the language server when I can TDDish it. (my own preference, but I think TDD is way more effective than the manual-test-loops many devs do).

**Review by themkat — COMMENTED** [review]
Your build fails after the latest commits 🙁 Also fails locally, which I tried just to make sure.

File: server/src/main/kotlin/org/javacs/kt/inlayhints/InlayHint.kt
Comment (ElamC) [reply server/src/main/kotlin/org/javacs/kt/inlayhints/InlayHint.kt:198]: I did add a test for this:

https://github.com/fwcd/kotlin-language-server/blob/c58f60054f5ff5f9b3ceb14a7b833095ea7f42ae/server/src/test/kotlin/org/javacs/kt/InlayHintTest.kt#L77

**Comment by ElamC** [comment]
> Your build fails after the latest commits 🙁 Also fails locally, which I tried just to make sure. Also seems like it stopped working in the editors (VSCode + Emacs)

My bad, tests fail it seems due to setting the defaults to false (not rendering hints and why it dosen't show up in your editor):
https://github.com/fwcd/kotlin-language-server/blob/2f031129911d62e34aaeca9d9973f5968bc34f61/server/src/main/kotlin/org/javacs/kt/Configuration.kt#L49

Is there a clean way to call config through tests? or would we prefer to set defaults as true i.e. leaving all hints all and having the option to disable them (tests will work that way too)

**Comment by themkat** [comment]
> Is there a clean way to call config through tests?

The language server test fixture creates a `languageServer` object. This `KotlinLanguageServer` object has an instance of config, and I'm pretty sure you can just set the config with that one:
https://github.com/fwcd/kotlin-language-server/blob/2f031129911d62e34aaeca9d9973f5968bc34f61/server/src/main/kotlin/org/javacs/kt/KotlinLanguageServer.kt#L26

**Review by themkat — APPROVED** [review]
Looks great to me, and works great 🙂 Awesome job! 🎉 

Will give @fwcd a day or two before merging in case he wants to have a look.

**Comment by themkat** [comment]
Let's merge it so people can start trying it! 😄 Good job @ElamC ! 🎉

**Pull request merged**
